stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,573 @@
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	// If nothing else fails, iomux service will be waiting here forever
+	// and be terminated by systemd only when the main application exits.
+	return <-c
+}
+
+// bufferLine buffers and queues a single line from a Reader to a multiplexer
+// If reading from src fails, it hard-fails and propagates the error back.
+func bufferLine(src io.Reader, c chan<- []byte, ec chan<- error) {
+	rd := bufio.NewReader(src)
+	for {
+		lineOut, err := rd.ReadBytes('\n')
+		if len(lineOut) == 0 && err != nil {
+			ec <- err
+		}
+		c <- lineOut
+	}
+}
+
+// acceptConn accepts a single client and queues it for further proxying
+// It is never canceled explicitly, as it is bound to the lifetime of the main process.
+func acceptConn(socket net.Listener, c chan<- net.Conn, stream string) {
+	for {
+		conn, err := socket.Accept()
+		if err == nil {
+			diag.Printf("Accepted new connection for %s\n", stream)
+			c <- conn
+		}
+	}
+}
+
+// proxyIO performs bi-directional byte-by-byte forwarding
+// TODO(lucab): this may become line-buffered and muxed to logs.
+// TODO(lucab): reset terminal state on new attach here
+func proxyIO(clients <-chan net.Conn, tty *os.File, ttyFailure chan<- error) {
+	ec := make(chan error)
+
+	// copyCloser copies remote<->tty.
+	// If copying from remote fails, we recovers and just close this connection.
+	// If copying from tty reaches EOF, it hard-fails and propagates the error up.
+	copyCloser := func(dst io.WriteCloser, src io.ReadCloser, recoverableErr bool) {
+		_, err := io.Copy(dst, src)
+		if recoverableErr {
+			src.Close()
+		} else if err == nil {
+			dst.Close()
+			ec <- io.EOF
+		}
+	}
+
+	for {
+		select {
+		case cl := <-clients:
+			go copyCloser(tty, cl, true)
+			go copyCloser(cl, tty, false)
+		case tf := <-ec:
+			ttyFailure <- tf
+			return
+		}
+	}
+}
+
+// muxInput accepts remote clients and multiplex input line from them
+func muxInput(clients <-chan net.Conn, stdin *os.File) {
+	for {
+		select {
+		case c := <-clients:
+			go bufferInput(c, stdin)
+		}
+	}
+}
+
+// bufferInput buffers and write a single line from a remote client to the local app
+func bufferInput(conn net.Conn, stdin *os.File) {
+	rd := bufio.NewReader(conn)
+	for {
+		lineIn, err := rd.ReadBytes('\n')
+		if err != nil {
+			conn.Close()
+			break
+		}
+		_, err = stdin.Write(lineIn)
+		if err != nil {
+			conn.Close()
+			break
+		}
+	}
+}
+
+// muxOutput receives remote clients and local log targets,
+// multiplexing output lines to them
+func muxOutput(streamLabel string, lines chan []byte, clients <-chan net.Conn, targets <-chan io.WriteCloser) {
+	var logs []io.WriteCloser
+	var conns []io.WriteCloser
+
+	for {
+		select {
+		// an incoming output line to multiplex
+		case l := <-lines:
+			// TODO(lucab): ordered non-blocking writes
+			for i := len(logs) - 1; i >= 0; i-- {
+				out := fmt.Sprintf("%s %s %s", time.Now().Format(time.RFC3339Nano), streamLabel, l)
```

Comment:
For hygiene we should dependency inject the formatting function having a `func(string) string` signature.